php/ext/google/protobuf2/names.c

```diff
@@ -0,0 +1,211 @@
+    sink->size = new_size;
+  }
+
+  memcpy(sink->ptr + sink->len, ptr, len);
+  sink->len += len;
+
+  return len;
+}
+
+static void stringsink_init(stringsink *sink) {
+  sink->size = 32;
+  sink->ptr = malloc(sink->size);
+  PBPHP_ASSERT(sink->ptr != NULL);
+  sink->len = 0;
+}
+
+static void stringsink_uninit(stringsink *sink) { free(sink->ptr); }
+
+/* def name -> classname ******************************************************/
+
+const char *const kReservedNames[] = {
+    "abstract",   "and",        "array",        "as",           "break",
+    "callable",   "case",       "catch",        "class",        "clone",
+    "const",      "continue",   "declare",      "default",      "die",
+    "do",         "echo",       "else",         "elseif",       "empty",
+    "enddeclare", "endfor",     "endforeach",   "endif",        "endswitch",
+    "endwhile",   "eval",       "exit",         "extends",      "final",
+    "for",        "foreach",    "function",     "global",       "goto",
+    "if",         "implements", "include",      "include_once", "instanceof",
+    "insteadof",  "interface",  "isset",        "list",         "namespace",
+    "new",        "or",         "print",        "private",      "protected",
+    "public",     "require",    "require_once", "return",       "static",
+    "switch",     "throw",      "trait",        "try",          "unset",
+    "use",        "var",        "while",        "xor",          "int",
+    "float",      "bool",       "string",       "true",         "false",
+    "null",       "void",       "iterable"};
+const int kReservedNamesSize = 73;
+
+bool is_reserved_name(const char* name) {
+  for (int i = 0; i < kReservedNamesSize; i++) {
+    if (strcmp(kReservedNames[i], name) == 0) {
+      return true;
+    }
+  }
+  return false;
+}
+
+static bool is_reserved(const char *segment, int length) {
+  bool result;
+  char* lower = malloc(length + 1);
+  memset(lower, 0, length + 1);
+  memcpy(lower, segment, length);
+  int i = 0;
+  while(lower[i]) {
+    lower[i] = (char)tolower(lower[i]);
+    i++;
+  }
+  lower[length] = 0;
+  result = is_reserved_name(lower);
+  free(lower);
+  return result;
+}
+
+static void fill_prefix(const char *segment, int length,
+                        const char *prefix_given,
+                        const char *package_name,
+                        stringsink *classname) {
+  if (prefix_given != NULL && strcmp(prefix_given, "") != 0) {
+    stringsink_string(classname, prefix_given, strlen(prefix_given));
+  } else {
+    if (is_reserved(segment, length)) {
+      if (package_name != NULL &&
+          strcmp("google.protobuf", package_name) == 0) {
+        stringsink_string(classname, "GPB", 3);
+      } else {
+        stringsink_string(classname, "PB", 2);
+      }
+    }
+  }
+}
+
+static void fill_segment(const char *segment, int length,
+                         stringsink *classname, bool use_camel) {
+  if (use_camel && (segment[0] < 'A' || segment[0] > 'Z')) {
```

Comment:
[Consistency] Since the code above uses `tolower`, would a `toupper` call be reasonable?
```
if (use_camel /* && islower(segment[0]) */) {
  char first = (char)toupper(segment[0]);
  ...
```

Reply:
I think I'm more inclined to change the other to *not* use `tolower()`, since its behavior is potentially affected by locales. I changed `tolower()` instead.

Reply:
Even better, IMO.